CASSANDRASC-135: Added additional JVM options to increase max open FD limit for MacOS

rwelgosh opened the pull request:
Patch by Raymond Welgosh; Reviewed by Francisco Guerrero, Yifan Cai for CASSANDRASC-135

Commit:
CASSANDRASC-135: Added additional JVM options to increase max open FD limit for MacOS

Patch by Raymond Welgosh; Reviewed by Francisco Guerrero, Yifan Cai for CASSANDRASC-135

Commit:
add changes

yifan-c commented on .java-version:
```diff
@@ -0,0 +1 @@
+graalvm64-11.0.20
```
we need to exclude this file and add a new entry to gitignore. 
The file is produced by `jenv`

frankgh commented on gradlew:
```diff
@@ -105,8 +105,10 @@ if [ "$cygwin" = "false" -a "$darwin" = "false" -a "$nonstop" = "false" ] ; then
 fi
 
 # For Darwin, add options to specify how the application appears in the dock
+# In addition, we want to disable the MaxFDLimit for MacOS, which is set to a lower limit than the actual system maximum.
```
```suggestion
# In addition, we want to increase the file descriptor limit to the MaxFDLimit in MacOS which, by default, is set to a lower limit than the actual system's maximum.
```

yifan-c replied:
To clarify, the JVM option `-XX:-MaxFDLimit` is to _disable_ the low `MaxFDLimit` set in JDK. By disabling, it can pick up the system FD limit, which is larger and proper. 

The suggested sentence looks incorrect.

frankgh replied:
I noticed there are places where we enable the flag, and other places where we disable the flag. Is that intentional?

Reading this documentation: https://docs.oracle.com/cd/E19683-01/806-7930/filestructure-11/index.html it looks like setting the flag i.e. doing `-XX:+MaxFDLimit` should `Bump the number of file descriptors to max`. Isn't this what we are trying to accomplish here?

yifan-c replied:
No. The max FD constant for macOS is low. It is `OPEN_MAX == 10240` https://github.com/openjdk/jdk/blob/dfd3da3f52480f68f653beb1e720691f8232ace7/src/hotspot/os/bsd/os_bsd.cpp#L1997-L2002

You want to _disable_ the max FD, so that it can pick up the system's limit.

bbotella commented on vertx-client-shaded/build.gradle:
```diff
@@ -83,6 +83,8 @@ shadowJar {
     }
 }
 
+test.dependsOn shadowJar
```
I'm curious. Why does it need the shadow jar now but didn't before?

yifan-c replied:
```
LibraryTest > Ensures that the shading is correct for the vertx-client-shaded project FAILED
```

It is to fix the flaky test, `LibraryTest`. The test requires shadow jar being built. However, when the test runs, the jar is empty. 

That being said, test already depends on shadowJar. It needs a different fix.

frankgh commented on src/test/java/org/apache/cassandra/sidecar/UlimitTest.java:
```diff
@@ -0,0 +1,44 @@
+        String line = reader.readLine();
+        long maxFD = Long.parseLong(line);
+        System.out.println(maxFD);
+        assertThat(maxFD).isGreaterThan(10240);
```
will this always be true? also, can we scope this test to macOS ?

frankgh commented on vertx-client-shaded/build.gradle:
```diff
@@ -83,6 +83,8 @@ shadowJar {
     }
 }
 
+test.dependsOn shadowJar
```
thanks for fixing this dependency! I have a small NIT:
```suggestion
tasks.named('test') {
    dependsOn 'shadowJar'
}
```

frankgh reviewed (commented):
Looks good. I have some minor small comments

yifan-c commented on client-common/build.gradle:
```diff
@@ -41,6 +44,9 @@ repositories {
 test {
     useJUnitPlatform()
+    if (Os.isFamily(Os.FAMILY_MAC)) {
+        jvmArgs "-XX:+MaxFDLimit"
```
disable. use `'-XX:-MaxFDLimit` instead.

Commit:
CASSANDRASC-135/trunk: fixed issues reported from PR Conversation

Switched the flag '-XX:+MaxFDLimit' to '-XX:-MaxFDLimit' and made applicable changes to comments.

Commit:
CASSANDRASC-135/trunk: Fixed CircleCI error

Syntax errors in build.gradle file fixed.

Commit:
Fixed issues with UlimitTest

Commit:
Disabled Ulimit Test

ulimit does not exist at the CI runtime so added a disabled annotation.

frankgh commented on src/test/java/org/apache/cassandra/sidecar/UlimitTest.java:
```diff
@@ -0,0 +1,54 @@
+ * Test to ensure that the max file descriptors is large enough to build Cassandra Sidecar
+ */
+@Disabled("ulimit does not exist at the CI runtime")
+public class UlimitTest
```
can we instead try annotating the test with `@EnabledOnOs({OS.MAC})`?

yifan-c replied:
Some annotations in junit5 are amazing!

Commit:
Fixed annotation on Ultimit test

Commit:
Deleted unused imports

frankgh reviewed (approved):
+1 Looks good to me
